topic_metadata.go

```diff
@@ -120,27 +135,160 @@ func (t *Topic) parsePartitions(value []byte) (PartitionList, error) {
 	return result, nil
 }
 
+// marshalPartitions turns a PartitionList into the JSON representation
+// to be stored in Zookeeper.
+func (t *Topic) marshalPartitions(partitions PartitionList) ([]byte, error) {
+	tm := topicMetadata{Version: 1, Partitions: make(map[string][]int32, len(partitions))}
+	for _, part := range partitions {
+		tm.Partitions[fmt.Sprintf("%d", part.ID)] = part.Replicas
+	}
+	return json.Marshal(tm)
+}
+
+// generatePartitionAssignments creates a partition list for a topic. The primary replica for
+// each partition is assigned in a round-robin fashion starting at a random broker.
+// Additional replicas are assigned to subsequent brokers to ensure there is no overlap
+func (t *Topic) generatePartitionAssignments(brokers []int32, partitionCount int, replicationFactor int) (PartitionList, error) {
+	if partitionCount <= 0 {
+		return nil, ErrInvalidPartitionCount
+	}
+	if replicationFactor <= 0 || len(brokers) < replicationFactor {
+		return nil, ErrInvalidReplicationFactor
+	}
+
+	result := make(PartitionList, partitionCount)
+
+	brokerCount := len(brokers)
+	brokerIdx := rand.Intn(brokerCount)
+
+	for p := 0; p < partitionCount; p++ {
+		partition := &Partition{topic: t, ID: int32(p), Replicas: make([]int32, replicationFactor)}
+
+		// assign primary to the current broker and all
+		// other replicas to subsequent brokers
+		for r := 0; r < replicationFactor; r++ {
+			partition.Replicas[r] = brokers[(brokerIdx+r)%brokerCount]
+		}
+
+		result[p] = partition
+		brokerIdx = (brokerIdx + 1) % brokerCount
+	}
```

Comment:
Maybe this is a simpler, and more random way to get a useful list of broker IDs:

``` go
brokerIndices = rand.Perm(len(brokers))[0:replicationFactor]
for r := 0; r < replicationFactor; r++ {
  partition.Replicas[r] = brokers[brokerIndices[r]]
}
```

Makes for simpler logic, and should give a even spread on a large amount of partitions. Feel free to disagree with this.